db_dtypes/__init__.py

```diff
@@ -346,35 +337,55 @@ def __sub__(self, other):
-        "and Python 3.8 in a future major release expected in Q4 2024. "
-        f"Your Python version is {sys_major}.{sys_minor}.{sys_micro}. We "
-        "recommend that you update soon to ensure ongoing support. For "
-        "more details, see: [Google Cloud Client Libraries Supported Python Versions policy](https://cloud.google.com/python/docs/supported-python-versions)",
-        PendingDeprecationWarning,
-    )
-
-
-if not JSONArray or not JSONDtype:
-    __all__ = [
+def _determine_all(json_array_type, json_dtype_type):
+    """Determines the list for __all__ based on JSON type availability."""
+    base_all = [
         "__version__",
         "DateArray",
         "DateDtype",
         "TimeArray",
         "TimeDtype",
     ]
-else:
-    __all__ = [
-        "__version__",
-        "DateArray",
-        "DateDtype",
-        "JSONDtype",
-        "JSONArray",
-        "JSONArrowType",
-        "TimeArray",
-        "TimeDtype",
-    ]
+    # Check if both JSON types are available (truthy)
+    if json_array_type and json_dtype_type:
+        # print("DEBUG: Condition FALSE, including JSON in __all__") # Keep if needed
+        return base_all + ["JSONDtype", "JSONArray", "JSONArrowType"]
+    else:
+        # print("DEBUG: Condition TRUE, excluding JSON from __all__") # Keep if needed
+        return base_all
+
+def _check_python_version():
+    """Checks the runtime Python version and issues a warning if needed."""
+    sys_major, sys_minor, sys_micro = _versions_helpers.extract_runtime_version()
+    if sys_major == 3 and sys_minor in (7, 8):
+        warnings.warn(
+            "The python-bigquery library as well as the python-db-dtypes-pandas library no "
+            "longer supports Python 3.7 and Python 3.8. "
+            f"Your Python version is {sys_major}.{sys_minor}.{sys_micro}. We "
+            "recommend that you update soon to ensure ongoing support. For "
+            "more details, see: [Google Cloud Client Libraries Supported Python Versions policy](https://cloud.google.com/python/docs/supported-python-versions)",
+            FutureWarning,
+            stacklevel=2, # Point warning to the caller of __init__
+        )
+
+
+# sys_major, sys_minor, sys_micro = _versions_helpers.extract_runtime_version()
+# if sys_major == 3 and sys_minor in (7, 8):
+#     warnings.warn(
+#         "The python-bigquery library as well as the python-db-dtypes-pandas library no "
+#         "longer supports Python 3.7 and Python 3.8. "
+#         f"Your Python version is {sys_major}.{sys_minor}.{sys_micro}. We "
+#         "recommend that you update soon to ensure ongoing support. For "
+#         "more details, see: [Google Cloud Client Libraries Supported Python Versions policy](https://cloud.google.com/python/docs/supported-python-versions)",
+#         FutureWarning,
+#     )
+
+# Perform the version check
+_check_python_version()
+
+# Assign __all__ by calling the function
+__all__ = _determine_all(JSONArray, JSONDtype)
```

Comment:
This seems a bit too "magic" to me, but I suppose it could make testing easier?

Reply:
This code was originally just in the file with no function def. We added _check_python_version...

To test against supported and unsupported versions we attempted to mock the version but timing became an issue, partly because dependencies also rely on version checks so our mock ended up preventing error-free installs of some dependencies.

We tried half a dozen ways to alter when the mock happened but none hit that sweet spot of reliably allowing the dependencies to do their thing and slipping in to allow our test to run.

Similarly, we ran into a mess of problems with import timing and mocking whether JSONArray JSONDtype did or did not import as expected and so ... 

putting them in functions made things much easier in the testing arena.

Reply:
Any hard feelings if we resolve this conversation?